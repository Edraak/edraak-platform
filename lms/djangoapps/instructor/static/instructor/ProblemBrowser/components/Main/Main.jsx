/* global gettext */
import { Button } from '@edx/paragon';
import { BlockBrowser } from 'BlockBrowser';
import * as PropTypes from 'prop-types';
import * as React from 'react';

export default class Main extends React.Component {
  constructor(props) {
    super(props);
    this.handleToggleDropdown = this.handleToggleDropdown.bind(this);
    this.state = {
      showDropdown: false,
    };
  }

  handleToggleDropdown() {
    this.props.fetchCourseBlocks(this.props.courseId, this.props.excludeBlockTypes);
    this.setState({ showDropdown: !this.state.showDropdown });
  }

  hideDropdown() {
    this.setState({ showDropdown: false });
  }

  render() {
    const { selectedBlock, onSelectBlock } = this.props;

    return (
      <div className="problem-browser">
        <Button onClick={this.handleToggleDropdown} label={gettext('Select a section or problem')} />
        <input type="text" name="problem-location" value={selectedBlock} disabled />
        {this.state.showDropdown &&
        <BlockBrowser onSelectBlock={(blockId) => {
          this.hideDropdown();
          onSelectBlock(blockId);
        }}
        />}
      </div>
    );
  }
}

Main.propTypes = {
  courseId: PropTypes.string.isRequired,
  excludeBlockTypes: PropTypes.arrayOf(PropTypes.string),
  fetchCourseBlocks: PropTypes.func.isRequired,
  onSelectBlock: PropTypes.func.isRequired,
  selectedBlock: PropTypes.string,
};

Main.defaultProps = {
  excludeBlockTypes: null,
  selectedBlock: null,
};
