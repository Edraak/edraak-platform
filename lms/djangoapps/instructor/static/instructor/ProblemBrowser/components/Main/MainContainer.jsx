import { fetchCourseBlocks, selectBlock } from 'BlockBrowser/data/actions/courseBlocks';
import { connect } from 'react-redux';

import Main from './Main';

const mapStateToProps = state => ({
  selectedBlock: state.selectedBlock,
});


const mapDispatchToProps = dispatch => ({
  onSelectBlock: blockId => dispatch(selectBlock(blockId)),
  fetchCourseBlocks: (courseId, excludeBlockTypes) =>
    dispatch(fetchCourseBlocks(courseId, excludeBlockTypes)),
});

const MainContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(Main);

export default MainContainer;
