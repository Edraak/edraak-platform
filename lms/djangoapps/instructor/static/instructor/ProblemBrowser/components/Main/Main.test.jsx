/* global jest,test,describe,expect */
import { Button } from '@edx/paragon';
import { BlockBrowser } from 'BlockBrowser';
import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import Main from './Main';

describe('ProblemBrowser Main component', () => {
  const courseId = 'testcourse';
  const excludedBlockTypes = [];

  test('render with basic parameters', () => {
    const component = renderer.create(
      <Main
        courseId={courseId}
        excludeBlockTypes={excludedBlockTypes}
        fetchCourseBlocks={jest.fn()}
        onSelectBlock={jest.fn()}
        selectedBlock={null}
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('render with selected block', () => {
    const component = renderer.create(
      <Main
        courseId={courseId}
        excludeBlockTypes={excludedBlockTypes}
        fetchCourseBlocks={jest.fn()}
        onSelectBlock={jest.fn()}
        selectedBlock={'some-selected-block'}
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('fetch course block on toggling dropdown', () => {
    const fetchCourseBlocksMock = jest.fn();
    const component = renderer.create(
      <Main
        courseId={courseId}
        excludeBlockTypes={excludedBlockTypes}
        fetchCourseBlocks={fetchCourseBlocksMock}
        onSelectBlock={jest.fn()}
        selectedBlock={'some-selected-block'}
      />,
    );
    const instance = component.getInstance();
    instance.handleToggleDropdown();
    expect(fetchCourseBlocksMock.mock.calls.length).toBe(1);
  });

  test('display dropdown on toggling dropdown', () => {
    const component = shallow(
      <Main
        courseId={courseId}
        excludeBlockTypes={excludedBlockTypes}
        fetchCourseBlocks={jest.fn()}
        onSelectBlock={jest.fn()}
        selectedBlock={'some-selected-block'}
      />,
    );
    component.find(Button).simulate('click');
    expect(component.find(BlockBrowser)).toBeTruthy();
  });
});
