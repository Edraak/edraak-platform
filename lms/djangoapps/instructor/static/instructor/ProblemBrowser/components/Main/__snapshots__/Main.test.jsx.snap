// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProblemBrowser Main component render with basic parameters 1`] = `
<div
  className="problem-browser"
>
  <button
    className="btn"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    Select a section or problem
  </button>
  <input
    disabled={true}
    name="problem-location"
    type="text"
    value={null}
  />
</div>
`;

exports[`ProblemBrowser Main component render with selected block 1`] = `
<div
  className="problem-browser"
>
  <button
    className="btn"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    Select a section or problem
  </button>
  <input
    disabled={true}
    name="problem-location"
    type="text"
    value="some-selected-block"
  />
</div>
`;
